proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,733 @@
+  that module.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Imports
+^^^^^^^
+
+Normal imports can only appear at the top level of a file. The target of a 
+normal import is a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the exported bindings of the target
+module are not brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope bound to that simple name instead.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The behaviour of the ``module`` export specifiers are changed as follows.
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and the listed in-scope bindings.
+- ``module M(..)`` exports ``M`` and all of its exported bindings. [#]_
+
+.. [#] See `Do we need additional syntax for re-exporting all bindings from a module?`_ for more discussion.
+
+Backwards compatibility extensions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Flat module exports
+^^^^^^^^^^^^^^^^^^^
+
+The extension ``FlatModuleExports`` is added and enabled by default. It has the effect
+that the ``module M`` export specifier has the behaviour given in [hs2010]_. [#]_
+
+.. [#] See `Flat module exports - reasoning`_ for reasoning.
+
+Import module merging
+^^^^^^^^^^^^^^^^^^^^^
+
+The extension ``ImportModuleMerging`` is added and enabled by default. It has the effect that
+when multiple modules are imported ``as`` the same name, we merge the modules. [#]_
+
+.. [#] See `Import module merging - reasoning`_ for reasoning. 
+   multiple modules are defined with the same name?`_ for discussion
+   about what the primary behaviour should be.
+
+Privileged local module access
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The extension ``PrivilegedLocalModuleAccess`` is added and enabled by default. It has the effect
+that for every file module ``M``, we add a new module declaration with the same name inside ``M``,
+which exports all and only the local declarations of ``M``. [#]_
+
+.. [#] See `Privileged local module access - reasoning`_ for reasoning.
+
+Additional features
+~~~~~~~~~~~~~~~~~~~
+
+The following extensions add features beyond existing Haskell, building upon the previous changes.
+
+Module imports
+^^^^^^^^^^^^^^
+
+We add the language extension ``ImportModule``, which introduces a a new
+syntax ``import module`` [#]_ for importing modules in other contexts.
```

Comment:
Where is this syntax allowed? Presumably, not intermixed with normal `import` statements. But is it allowed in `let` and `where` clauses?

Reply:
I guess the reason not to allow them in the initial block of `import` statements is to not break scanners that make assumptions about the structure of modules? I do think it would be good to allow them to be at the file-module level. But we could ban them in the initial block of `import`s.

Initially I had thought that we could have these only at the module level, but allowing them in other scopes would be very natural too. What about local module declarations in `let` and `where` clauses too?